### ISSUE TITLE
Better support for using substrate as a library

### ISSUE THREAD

**Issue by olanod** [paritytech/substrate#6885]
So far seems the development of Substrate chains is very focused on starting projects from scratch following the template node and then just fiddling with the runtime modules but it would be nice if there is better support in the form of documentation, examples and stable APIs on how to create our own nodes that can be embedded in an existing code base and interacting with the runtime(e.g. submitting transactions) on a library level instead of external JSON RPC calls.

**Comment by dvc94ch** [comment]
I think what you're looking for is https://github.com/paritytech/substrate-subxt

**Comment by kianenigma** [comment]
> I think what you're looking for is https://github.com/paritytech/substrate-subxt

> interacting with the runtime(e.g. submitting transactions) _on a library level_ **instead** of _external JSON RPC_ calls.

Based on my ^^ I interpreted the issue exactly the opposite, but either way sub-xt is a good tool to point out. 

--- 

What I understood was that you are more interested in using non-runtime parts of substrate (client, primitives) to build custom clients. This is indeed planned and essential for substrate's long term success but addmitedly it is not as mature and well documented as the runtime side. 

Can you explain more the type of tools that you want to build and the type of libraries that you expect substrate to provide? 

cc @gnunicorn

**Comment by dvc94ch** [comment]
subxt can use a light client. we can even drop in a full node, be use this for all our automated client testing.

**Comment by kianenigma** [comment]
then I guess what you mean is that sub-xt is a good _example_ of how to use substrate as a library? I thought that it is the _means_ to to so 😄.

**Comment by dvc94ch** [comment]
I'm confused. I guess subxt is a library to build substrate clients, and examples of how to use it would be projects using subxt.

**Comment by olanod** [comment]
> What I understood was that you are more interested in using non-runtime parts of substrate (client, primitives) to build custom clients. This is indeed planned and essential for substrate's long term success but addmitedly it is not as mature and well documented as the runtime side.

Indeed, what I'm after is a custom client with an application specific transport and protocol, other use case might be no over the network communication at all?(e.g. an electron-like all included node with UI submitting transactions to the runtime directly). Subxt looks like a good start, it can be the higher level API to interact with the runtime but ideally it can also work with an instance of the runtime perhaps? and submit extrinsics directly without the json rpc client-server layer?

**Comment by gnunicorn** [comment]
> Indeed, what I'm after is a custom client with an application specific transport and protocol, other use case might be no over the network communication at all?(e.g. an electron-like all included node with UI submitting transactions to the runtime directly)

If you look at `client/service` you'll notice that all components of substrate are actually (mostly standalone) libraries and the service crate primarily fiddles them together to a client (with the particular purpose to be easily extensible from the node-templates, hence it rather gigantic type definitions). You could use the result of that service within any other rust-lib to do whatever you like – afaik, this is how the raddicle people are building their oscoin on-top (see https://github.com/radicle-dev/radicle-registry ). The API just isn't very oriented through that usage and exposed functionality doesn't include transaction transmission or alike.

However, all it does can be done within any library and be used directly – there are some background task management you'd also have to do though – and then you can talk to the component yourself. Some of these components would require more specifically tied types and traits to be implemented on outside components, but all that should be available from `primitives/` – be aware though that many things within these substrate components are super generic to allow for the generic upgrade mechanism to work. Through doing this, one could technically build an in-process client which doesn't use libp2p at all for example, but receives messages through a completely separate channel. 

This is all possible today, but rather complex to do. While we intend to allow for easier integration, until this is there (and this is probably still a while to go) the recommended path is to transmit transactions through RPC to a running and connected node.
